test/compress/async.js

```diff
@@ -140,6 +140,182 @@ dont_inline: {
+        (async function() {
+            async function f() {
+                await 42;
+            }
+            return await f();
+        })();
+        console.log("PASS");
+    }
+    expect: {
+        (async function() {
+            await 42;
+        })();
+        console.log("PASS");
+    }
+    expect_stdout: "PASS"
+    node_version: ">=8"
+}
+
+inline_await_2: {
+    options = {
+        inline: true,
+        reduce_vars: true,
+        unused: true,
+    }
+    input: {
+        (async function() {
+            async function f(a) {
+                await a;
+            }
+            return await f(console);
+        })();
+        console.log("PASS");
+    }
+    expect: {
+        (async function() {
+            return await void await console;
+        })();
+        console.log("PASS");
+    }
+    expect_stdout: "PASS"
+    node_version: ">=8"
+}
+
+inline_await_2_trim: {
+    options = {
+        if_return: true,
+        inline: true,
+        reduce_vars: true,
+        side_effects: true,
+        unused: true,
+    }
+    input: {
+        (async function() {
+            async function f(a) {
+                await a;
+            }
+            return await f(console);
+        })();
+        console.log("PASS");
+    }
+    expect: {
+        (async function() {
+            await console;
+        })();
+        console.log("PASS");
+    }
+    expect_stdout: "PASS"
+    node_version: ">=8"
+}
+
+inline_await_3: {
+    options = {
+        inline: true,
+        reduce_vars: true,
+        unused: true,
+    }
+    input: {
+        (async function() {
+            async function f(a, b) {
+                return await b(a);
+            }
+            return await f("PASS", console.log);
+        })();
+    }
+    expect: {
+        (async function() {
+            return await (a = "PASS", b = console.log, await b(a));
+            var a, b;
+        })();
+    }
+    expect_stdout: "PASS"
+    node_version: ">=8"
+}
+
+inline_await_3_trim: {
+    options = {
+        inline: true,
+        reduce_vars: true,
+        sequences: true,
+        side_effects: true,
+        unused: true,
+    }
+    input: {
+        (async function() {
+            async function f(a, b) {
+                return await b(a);
+            }
+            return await f("PASS", console.log);
+        })();
+    }
+    expect: {
+        (async function() {
+            return a = "PASS", b = console.log, await b(a);
+            var a, b;
+        })();
+    }
+    expect_stdout: "PASS"
+    node_version: ">=8"
+}
+
+await_unary: {
+    options = {
+        side_effects: true,
+    }
+    input: {
+        (async function() {
+            console.log("PASS");
+            await +[];
+            console.log("FAIL");
+        })();
+    }
+    expect: {
+        (async function() {
+            console.log("PASS");
+            await +[];
+            console.log("FAIL");
+        })();
+    }
+    expect_stdout: "PASS"
```

Comment:
This is a bit misleading - it's only the stdout output due to a shortcoming of the compress test framework - before the next tick. It should output `"PASS\nFAIL"`. It would be nice to support true async tests - whether using `setTimeout`, generators/yield, or async/await. Might have to mark the test to fork and exec to do it however. Perhaps if there's an async function it could automagically have this behavior.

Reply:
The test framework could be changed to `await` every test result, but it could hang on a bad result.

Reply:
`vm` is genuinely unhelpful for anything asynchronous
https://nodejs.org/docs/latest/api/vm.html#vm_timeout_interactions_with_asynchronous_tasks_and_promises

I've even tried that `microtaskMode` thing with the latest version, and... please don't ask me how I feel.

Reply:
> The test framework could be changed to `await` every test result, but it could hang on a bad result.

No need to use any fancy bolt-on features &minus; we can simply attach to `.then()` and operate well within the limits of JavaScript.

The problem is that the queue runs on the global namespace, not the one provided by `vm.runInThisContext()` a.k.a. escaped the sandbox, so there's no way to meaningfully capture the output afterwards.

The only solution if I am paid a million bucks _and_ a gun to my head is to use `child_process.execSync()` or something along those lines, but let's just say none of those conditions have been met so... 😏

Reply:
> This is a bit misleading

Okay coming a long way back to this part 😅 &minus; would the following be less misleading?
```js
var a;
(async function() {
    a = "PASS";
    await +[];
    a = "FAIL";
})();
console.log(a);
```